### PR TITLE
K8s/Snapshots: Add dashboardsnapshot api group

### PR DESCRIPTION
**Pull request by ryantxu**
This exposes snapshots as k8s resources and implements existing create/delete api within the apiserver.

This requires running the feature flags:
```
[feature_toggles]
grafanaAPIServerWithExperimentalAPIs = true
# changes the UI to use K8s
kubernetesSnapshots = true
```

image

Why "dashsnap" and not "snapshots"?  -- "dashboardsnapshots" seems too long and wordy, "snapshots" is too general in the scope of grafana labs (we may want that for something else!)

NOTE: the apiserver routes do not (yet) work with anonymous users.  We can fix that in a follow up PR.

### CODE REVIEW

File: pkg/apis/snapshots/v0alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type DashboardSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Info SnapshotInfo `json:"info"`
+
+	// The raw dashboard (??? openapi ???)
+	Dashbaord any `json:"dashboard"`
```
Comment (torkelo) [inline pkg/apis/snapshots/v0alpha1/types.go:16]:
```suggestion
	Dashboard any `json:"dashboard"`
```

:smile:

Comment (ryantxu) [reply pkg/apis/snapshots/v0alpha1/types.go:16]: 🤕 i have never made a typo in my whole life

File: pkg/registry/apis/dashsnap/register.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+			},
+		},
+	}
+}
+
+func (b *SnapshotsAPIBuilder) GetAuthorizer() authorizer.Authorizer {
+	return authorizer.AuthorizerFunc(
+		func(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+			// TODO -- something more restrictive?
+			return authorizer.DecisionAllow, "", err
```
Comment (ryantxu) [inline pkg/registry/apis/dashsnap/register.go:257]: TODO... this is the biggest open question now.  We must audit that the existing permissions model still holds up.

File: pkg/api/dashboard_snapshot.go [hunk 3]
```diff
@@ -357,8 +167,12 @@ func (hs *HTTPServer) DeleteDashboardSnapshot(c *contextmodel.ReqContext) respon
 		return response.Error(http.StatusNotFound, "Failed to get dashboard snapshot", nil)
 	}
 
+	if queryResult.OrgID != c.OrgID {
+		return response.Error(http.StatusUnauthorized, "OrgID mismatch", nil)
```
Comment (ryantxu) [inline pkg/api/dashboard_snapshot.go:171]: ^^ fixes minor bug where users can delete snapshots across org boundaries

**Comment by ryantxu** [comment]
@evictorero can you take a look?  This does not change any existing behavior, but adds an additional endpoint (that does not yet support annon access)

NOTE: this requires devmode and:
```
[feature_toggles]
grafanaAPIServerWithExperimentalAPIs = true
```
so it will not be used in production yet

**Review by evictorero — APPROVED** [review]
I've tested it (both, the legacy api and the new one) and works as expected.
I left some nit comments.
I think we can merge it and continue working on it but it is a great start!

File: pkg/api/dashboard_snapshot.go [hunk 4]
```diff
@@ -357,8 +167,13 @@ func (hs *HTTPServer) DeleteDashboardSnapshot(c *contextmodel.ReqContext) respon
 		return response.Error(http.StatusNotFound, "Failed to get dashboard snapshot", nil)
 	}
 
+	// TODO: introduce orgID check
```
Comment (ryantxu) [inline pkg/api/dashboard_snapshot.go:170]: @evictorero -- lets add this in a follow up PR, the tests all pass locally, but not in in CI so I think there must be a difference in the org setup.  The fix will be subtle enough that it should not be included here

File: pkg/registry/apis/dashboardsnapshot/conversions.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+	dashboardsnapshot "github.com/grafana/grafana/pkg/apis/dashboardsnapshot/v0alpha1"
+	"github.com/grafana/grafana/pkg/services/dashboardsnapshots"
+	"github.com/grafana/grafana/pkg/services/grafana-apiserver/endpoints/request"
+	"github.com/grafana/grafana/pkg/services/grafana-apiserver/utils"
+)
+
+func convertDTOToSnapshot(v *dashboardsnapshots.DashboardSnapshotDTO, namespacer request.NamespaceMapper) *dashboardsnapshot.DashboardSnapshot {
+	expires := v.Expires.UnixMilli()
+	if v.Expires.After(time.Date(2070, time.January, 0, 0, 0, 0, 0, time.UTC)) {
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/registry/apis/dashboardsnapshot/conversions.go:17]: If all goes well, this code will be deleted in the next 6 months anyway so we should be safe :)

File: pkg/setting/setting.go [hunk 6]
```diff
@@ -359,12 +359,14 @@ type Cfg struct {
...
-	ExternalSnapshotName  string
-	ExternalEnabled       bool
+	SnapshotEnabled      bool
+	ExternalSnapshotUrl  string
+	ExternalSnapshotName string
+	ExternalEnabled      bool
+	// Deprecated: setting this to false adds deprecation warnings at runtime
 	SnapShotRemoveExpired bool
 
+	// Only used in https://snapshots.raintank.io/
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/setting/setting.go:369]: Could be true -- but I really want to reduce the scope of possible deployments we advertise/support/encourage.   I strongly feel this public server should be a (simple) standalone service and we remove it from core grafana.

File: pkg/registry/apis/dashboardsnapshot/register.go [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+						wrap.JsonApiErr(http.StatusBadRequest, "error getting options", err)
+						return
+					}
+
+					// Use the existing snapshot service
+					dashboardsnapshots.CreateDashboardSnapshot(wrap, opts.Spec, b.service)
+				},
+			},
+			{
+				Path: prefix + "/delete/{deleteKey}",
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/registry/apis/dashboardsnapshot/register.go:271]: Keeping it here so we can remove it from `/api` and just rewrite the commands for backwards compatibility.

**Pull request merged**